[RDY] Rework hordes hearing signal calculation

Firestorm01X2 opened the pull request:
Currently wander spawn (aka hordes) is hardcore mode for hardcore players. Actually even players who play Cataclysm for long time and who want to increase difficulty tend to avoid wander spawn.

The main reason of that - horde can hear gunshot from insane distance and have tricky and unfinished spawn and moving algorithms.

This PR tries to fix insane hearing distance and rework hearing formula for hordes.

Here what we have new:
* New method of calculation that calculates overmap distance hearing for hordes "naturally". Just like if normal volume in tiles reach it. With certain modification;
* New formula that normalizes hearing sounds to overmap without coefficient. It can be easily tweaked further. Actually it scales sound interval like [X1,Y1] to overmap interval [X2,Y2] (http://stackoverflow.com/questions/929103/convert-a-number-range-to-another-range-maintaining-ratio) (not in use now. may be useful in future) (removed from code for now. upper is better); 
* Reduced hearing for hordes. Now city hordes should not hear you if you are deep in wilderness;
* Hordes will hear only really loud sound that can leave reality bubble;
* Hordes can actually hear you underground and above ground. However if you stay under ground then ground cover sound partially. The lower you are - less sound you produce for hordes.


Discussion is here:
http://smf.cataclysmdda.com/index.php?topic=13632.0

Firestorm01X2 commented:
Currently  sound interval like [0,180] scaled to [0,8];

Firestorm01X2 commented:
How much loudnsess at max player can produce? 200 or even potentially more?

Coolthulhu commented:
Excessive defines make the code harder to read. They don't help here at all.

If you want all of it to be easy to read, configure and reuse, just wrap the "volume transformation" into a function that takes base noise and outputs range.

Firestorm01X2 commented:
Wraped extra code in functions. 
Also alternate simple solution added. Just divide volume/30+1 .30 - is overmap tile. Pretty straightforward and logical solution. This way is connected with normal volume calculation.
New method still in place. But maybe more simplier and natural method should be preffered?

Coolthulhu commented on src/sounds.cpp:
```diff
@@ -160,9 +161,48 @@ static std::vector<centroid> cluster_sounds( std::vector<std::pair<tripoint, int
+    int min_vol_for_hordes = 61;
+    int max_vol_for_hordes = 200;
+    int min_sig_power_for_hordes = 4;
+    int max_sig_power_for_hordes = 8;
```
Those names are giant, which makes it pretty hard to read the function.
It would be better if you shortened them to shortest readable name and added a comment if it's ambiguous.

For example `min_signal`, `max_vol`, `vol_signal_coeff` etc.

Coolthulhu commented on src/sounds.cpp:
```diff
@@ -160,9 +161,48 @@ static std::vector<centroid> cluster_sounds( std::vector<std::pair<tripoint, int
 void sounds::process_sounds()
 {
-    std::vector<centroid> sound_clusters = cluster_sounds( recent_sounds );
+	std::vector<centroid> sound_clusters = cluster_sounds( recent_sounds );
```
Looks like your IDE changed spaces here to tabs.
You'll need to change all tabs to 4 spaces each. Most IDEs support automatic tab->spaces.

Firestorm01X2 commented:
* Removed tabs
* Refactored vars name
* Limit to horde hearing correcly set to reality bubble (22*3)
* Changes simple version of algorithm to match overmap tile size

Now there are 2 versions of calculation algorithm. One- with normalization. Easily tweakable to any requirement. It just rescale volume to overmap tiles.

Second- simple and clear. It is just divide sound volume to overmap size in  player tiles. So volume/22 gives as horde signal. Sound levels are manageable. Shotgun shot shoud be heard from 8 overmap tiles.

Firestorm01X2 commented:
What is the max volume that potentially can be produced ingame?

Coolthulhu commented:
> Limit to horde hearing correctly set to reality bubble (22*3)

It's 24 per overmap tile. However reality bubble isn't exactly 3 overmap tiles, it's between 5 and 6 submaps (2.5 to 3 overmap tiles), because it isn't actually centered on the player, but on the submap containing the player.

> What is the max volume that potentially can be produced ingame?

There is no real limit. Unless I'm missing something, ANFO charge explosion should produce 4500 units of sound volume.

> Now there are 2 versions of calculation algorithm.

The simple one looks good enough. We don't need precision here, but readability is always a plus.

Coolthulhu commented on src/sounds.cpp:
```diff
@@ -160,6 +161,45 @@ static std::vector<centroid> cluster_sounds( std::vector<std::pair<tripoint, int
+{
+    int min_vol_cap = 67;//Hordes can't hear volume lower than this
+    int per_level_down_coef= 2;//Coeffficient for volume reduction undeground
+    int coef_for_hordes = 22;//Divider coefficent for hordes
```
Those two lower comments don't really add much.

Would be better like
```
int undeground_div = 2; // Underground noise is divided by this times number of z-levels below 0
int overmap_tile = 2 * SEEX; // Signal is in overmap tiles, volume in map tiles
```

Coolthulhu commented:
That underground noise thing needs to get the z-level of noise source, not global z-level.

To do that, you should pass it to the signal function from `process_sounds`. Currently it will stop signaling hordes if the player gets underground, even if the noise sources (NPC with a shotgun, turrets etc.) are on the surface.

Coolthulhu commented:
Needs some styling:
* Spaces within parentheses: `if( x )`, not `if( x)`, `if(x)` or `if(x )`
* Tabs to spaces (there are still some tabs in the diff)
* Spaces around operators: `int x = a`, not `int x=a`

Voskhod-1 commented:
In truth, the reason I don't like hordes is because they only count sounds within the reality-bubble as worthy of notice. They never, for instance, migrate towards military outposts or roadblocks that have anti-material turrets that are sure to be firing even in the player's absence. Nor do they appear to ever notice minefields that are sure to be stepped on by the odd wildlife or two, even when the player isn't around to witness this happening.

Really though, It'd probably be pretty hard to code sounds happening outside of the reality bubble. But, just a suggestion.

Firestorm01X2 commented:
Also is sound unit (vol in my case) in tiles exactly?
An abot that explosion of 4500 - top capped is needed in this case.

Coolthulhu commented:
> Also is sound unit (vol in my case) in tiles exactly?

Yes.

> An abot that explosion of 4500 - top capped is needed in this case.

Cap around 150 would be fine. Alternatively, a soft cap, with everything above the cap going through some hardcore stepdown, like: `if( sig > 8 ) sig = 8 + log( sig - 8 )`.

Firestorm01X2 commented:
* z-level now getting from sound position
* Added soft cap to get_signal_for_hordes_simple function
* Divider in get_signal_for_hordes_simple  now equal to tiles in overmap tiles 2 * SEEX
* Refactored code
* Added bunch of coefficients to correct signal value
* Both funtions balanced to: 
      glock 19 shot - 5  overmaptiles
      autoshotgun shot  - 6  overmaptiles
      amininuke explosive  - 11-12  overmaptiles

pisskop commented:
Since the thread was used as justification of this, Im weighing in.

5 tiles for s pistol is ridiculously small.  thats a 5 minute walk.  Ive suggests the rule of thumb "sound/6" as a guide.  Thats double for the pistol, and easily conservative in what it would draw.

Firestorm01X2 commented:
Ok. Let's remove set my special coefficient to zero.

Firestorm01X2 commented:
> "sound/6"
Give you 27 overmap tiles with glock 19 and JHP ammo. It sound 160.

Firestorm01X2 commented:
Now:
![2016-11-12 21-54-18](https://cloud.githubusercontent.com/assets/17512620/20240175/68229f64-a92b-11e6-82c1-449349ae5b03.png)
![2016-11-12 21-55-26](https://cloud.githubusercontent.com/assets/17512620/20240176/682427da-a92b-11e6-9f7a-f0c894c390a8.png)
![2016-11-12 21-56-10_m](https://cloud.githubusercontent.com/assets/17512620/20240177/68246ed4-a92b-11e6-9a03-32799c3aeae3.png)

Firestorm01X2 commented:
Also you can see that actual sound from glock and autoshotgun almost similar.

hmstanley commented:
Somewhat of an off topic question, Firestorm, what font are you using in game?  I'd like to implement it.

Firestorm01X2 commented:
>Somewhat of an off topic question, Firestorm, what font are you using in game? I'd like to implement it. 

Looks like it is Monospace Regular. It bundled with Linux Mint and set as default to its console.

Firestorm01X2 commented:
Currently used "Simple" version of algorithm that calculates signal using divider  that equals 2*SEEX.

Firestorm01X2 commented:
Ha. It is not ready yet. Looks like signal to hordes connected to interest values.
src/overmap.cpp
void overmap::signal_hordes( const tripoint &p, const int sig_power)
{
...
2926: const int d_inter = (sig_power - dist) * 5;
...
}

And again that "5" modifier.

Firestorm01X2 commented:
Question, What should I cgange into this(overmap.cpp):
`/**
* @param sig_power - power of signal or max distantion for reaction of zombies
*/
void overmap::signal_hordes( const tripoint &p, const int sig_power)
{
    for( auto &elem : zg ) {
        mongroup &mg = elem.second;
        if( !mg.horde ) {
            continue;
        }
            const int dist = rl_dist( p, mg.pos );
            if( sig_power <= dist ) {
                continue;
            }
            // TODO: base this in monster attributes, foremost GOODHEARING.
            const int d_inter = (sig_power - dist) * 5;
            const int roll = rng( 0, mg.interest );
            if( roll < d_inter ) {
                // TODO: Z coord for mongroup targets
                const int targ_dist = rl_dist( p, mg.target );
                // TODO: Base this on targ_dist:dist ratio.
                if (targ_dist < 5) {
                    mg.set_target( (mg.target.x + p.x) / 2, (mg.target.y + p.y) / 2 );
                    mg.inc_interest( d_inter );
                } else {
                    mg.set_target( p.x, p.y );
                    mg.set_interest( d_inter );
                }
            }
    }
}`

To correctly support signal between [4,10].
I tested hod did new calculation attract hordes, And find that hordes almost do not react to sound. And this is not because distance (I think).

Firestorm01X2 commented:
I've got the problem.
const int dist = rl_dist( p, mg.pos );
It is calculate distance in strange way. I see distance 3-4 on overmap. However  rl_dist returns different and much bigger distance around 7.

Firestorm01X2 commented:
Looks like I figured that out. rl_dist returns result in submaps. 1 overmap = 2 submaps (or at least approximately). So now I calculate singal spower in submaps. 
Currently hordes react to shot from 6-7 tiles. Also it is matching rule: if you see horde - than it may  then it will hear your shoot.

Also added balance coefficients for "Simple" function. It will help to balance it later.

pisskop commented:
Guns dont really need a rework per se.

A mile is probably equatable to 12 tiles or so.


What should be fixed is how hordes draw hordes to themselves -but only within the reality bubble i.e your location

Firestorm01X2 commented:
Tests show that with 7 overmap tiles you can still draw horde from city while shooting in starting shelter. It is even without no signal modifying at all - without reducing it with coefficients or increasing it. It is surprisingly good results. Almost like someone planning extension of reality bubble.

Firestorm01X2 commented:
Take notes that int get_signal_for_hordes_using_normalization is in use now.
I use int get_signal_for_hordes( const centroid &centr ). Other method may be useful for future balancing.

Ok. I think that it is done.

pisskop, look at the get_signal_for_hordes( const centroid. There are coeffcients for balancing, You can easily tweak them in your PR to achieve 10  overmap tiles for pistol shot. Just set:
` float sig_coef = 1.3;`
instead of 1

Ready.

Firestorm01X2 commented:
Removed unused method. Signal value connected with horde interest value. So signal is not so independent.

Also I can implement that altrnative method later but only if it will be needed. For now usual method of calculation produce fluid results.

Ready.

Coolthulhu commented on src/overmap.cpp:
```diff
@@ -2919,22 +2920,25 @@ void overmap::signal_hordes( const tripoint &p, const int sig_power)
-            const int d_inter = (sig_power - dist) * 5;
+            const int d_inter = (sig_power + 1 - dist) * SEEX;
             const int roll = rng( 0, mg.interest );
+            add_msg( m_debug, "horde roll % d_inter %d dist %d", roll, d_inter, dist);
```
That format string looks wrong.

Coolthulhu commented on src/sounds.cpp:
```diff
@@ -160,6 +161,34 @@ static std::vector<centroid> cluster_sounds( std::vector<std::pair<tripoint, int
+        //Capping minimum horde hearing signal
+        sig_power = ( ( sig_power < min_sig_cap ) ? min_sig_cap : sig_power );
+        //Capping extremely high signal to hordes
+        sig_power = ( ( sig_power > max_sig_cap ) ? max_sig_cap : sig_power );
```
It's clearer with `std::min` and `std::max`, like `sig_power = std::min( sig_power, max_sig_cap )`.

Also, the comment above suggests that the cap is soft (ie. that sig can sometimes go above it). Does it still do that? If not, the comment should be changed.

Coolthulhu commented on src/sounds.cpp:
```diff
@@ -160,6 +161,34 @@ static std::vector<centroid> cluster_sounds( std::vector<std::pair<tripoint, int
+    const int max_sig_cap = 26;//Signal for hordes beyond this cap will be reduced heavily
+    //Balancing coefs:
+    const float sig_coef = 1;// To make player life easier - make it less than one. To make harder - more than  one
+    const float sig_modifier = 0;// To make player life easier - make it less than zero. To make harder - more than zero
```
This doesn't look useful for anything.

Coolthulhu commented:
And again: formatting. Spaces in `if`s and no tabs.

Firestorm01X2 commented:
Now it is really ready.

pisskop commented:
this means most basements and science migoa wont be heard by hordes.


does this affect player hearing or hearing spam

Firestorm01X2 commented:
>this means most basements and science migoa wont be heard by hordes.
About basements - before this PR hordes won't be able to hear sound from basement at all.

Coolthulhu commented:
> About basements - before this PR hordes won't be able to hear sound from basement at all or from any level, different from surface at all.

Actually worse than that - old horde hearing would mean that hordes could hear a mi-go in the lab just fine, but only as long as the player stayed on 0 z-level.

kevingranade commented:
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/request-to-developers/22170/1
